elastic/kibana
# PoC: Derive OAS 3 from Kibana router

## Summary

PoC to explore whether we can generate good/high quality OAS spec from our new versioned router. The general idea is to bring the versioned router interface as inline as possible with OAS 3 so that we can map from router declarations as easily as possible.

This PoC tests the following runtime validation libs:
* `@kbn/config-schema` - core provided runtime validation lib powered by `joi` (using `joi-to-json` for JSONSchema)
* [`zod`](https://github.com/colinhacks/zod) ([see docs](https://zod.dev/)), a well supported validation library that can be used to [generate JSONSchema](https://github.com/StefanTerdell/zod-to-json-schema).

This is a departure from how our current OAS spec is being created and is much more developer-centric. Developers own the output (and quality) of OAS for public routes.

See full OAS output [here](https://gist.github.com/jloleysens/ba3f047706fe471fede8e3ba4ca82340). Note: this contains OAS for ALL registered routes so it is really big.

## How to test deriving OAS for your endpoints

* Zod or kbn/config-schema are supported libs
* Use the new `/api/oas` endpoint. Issue a request like: `curl -uelastic:changeme http://localhost:5601/<base-path-thingy>/api/oas` (optionally format with `| jq` for legibility)
* (optional, but recommended) You can filter for paths starting with a certain value using the queryParam `pathStartsWith`. For example `/api/alerting/rule/` will generate alerting rules OAS (make sure to correctly escape).


### For versioned routes

1. Add body and response schemas
2. Start Kibana
3. See result using `/api/oas?pathStartsWith=/my/api`

### For non-versioned routes

1. Ensure body schema is defined
2. Use the new `responses` key to describe responses for your endpoint
4. Start Kibana
5. See result using `/api/oas?pathStartsWith=/my/api`

## Example

<details>
<summary>Given router declaration like</summary>

```ts
router
  .get({
    access: 'public',
    path: '/api/cool/endpoint/{id}',
  })
  .addVersion(
    {
      version: '2023-01-01',
      description: 'My cool route',
      validate: {
        request: {
          params: z.object({ id: z.string() }),
        },
        response: {
          200: {
            body: z.object({ foo: z.string().max(5) }),
          },
          400: {
            body: z.object({ error: z.string() }),
          },
        },
      },
    },
    (ctx, req, res) => res.ok()
  )
```

</details>

<details>

<summary>Generate OAS 3 like</summary>

```json
{
  "openapi": "3.0.0",
  "info": {
    "title": "PoC",
    "version": "1.0.0"
  },
  "servers": [
    {
      "url": "http://localhost:5601"
    }
  ],
  "paths": {
    "/api/cool/endpoint/{id}": {
      "get": {
        "responses": {
          "200": {
            "description": "No description",
            "content": {
              "application/json; Elastic-Api-Version=2023-01-01": {
                "schema": {
                  "type": "object",
                  "properties": {
                    "foo": {
                      "type": "string",
                      "maxLength": 5
                    }
                  },
                  "required": [
                    "foo"
                  ],
                  "additionalProperties": false
                }
              }
            }
          },
          "400": {
            "description": "No description",
            "content": {
              "application/json; Elastic-Api-Version=2023-01-01": {
                "schema": {
                  "type": "object",
                  "properties": {
                    "error": {
                      "type": "string"
                    }
                  },
                  "required": [
                    "error"
                  ],
                  "additionalProperties": false
                }
              }
            }
          }
        },
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string"
            }
          }
        ]
      }
    }
  }
}
```

</details>

## TODOs

The script for doing this will cover

- [x] Request body
- [x] Responses
- [x] URL parameters
- [x] Query parameters (still need to test)
- [ ] What else...?

## Benefits

1️⃣ **Single schema declaration** (source of truth):

* Runtime validations
* TypeScript
* OAS spec (new)

```typescript
export const coolSchema = z.object({ foo: z.string().min(5) }); 
export type CoolSchema = z.infer<typeof anyRecord>;

/**
If used in router declaration will result in:
{
  ...
  "foo": {
    "type": "string",
    "minLength": 5
  }
*/
```

## Drawbacks

* ~Need to refactor existing runtime schemas to use `@kbn/zod`, this drawback can be mitigated to work with `@kbn/config-schema` if we use something like https://github.com/lightsofapollo/joi-to-json-schema, but this was not investigate further.~ Refactored to support joi
* Need to maintain OAS generation script
* May result in lower quality OAS (can largely be mitigated bc we control the versioned router interface)

## Out of scope (future tasks)

* Not all OAS 3 capabilities are covered, yet. Specifically
  * Examples
  * Summary
* Figuring out how all paths for Kibana will ultimately be combined or split up for search, observability and security projects
* Special PR/CR checks that can be used to update or check OAS
* Only support for application/json requests and responses ATM, we need to support any response content type
* Support for spaces (we need to generate a version of all paths with a space prefix)

## Screenshots

<img width="1359" alt="Screenshot 2023-04-21 at 16 29 45" src="https://user-images.githubusercontent.com/8155004/233662129-0757d04a-76d5-4e0b-9238-26cc645f0412.png">

## Related
- [Example of using versioned router](https://docs.elastic.dev/kibana-dev-docs/versioning-http-apis#4-adhere-to-the-http-versioning-specification)
